odiglet/pkg/kube/runtime_details/instrumentationconfigs_controller.go

```diff
@@ -0,0 +1,136 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type instrumentationConfigPredicate struct{}
+
+func (p *instrumentationConfigPredicate) Create(e event.CreateEvent) bool {
+	obj, ok := e.Object.(*odigosv1.InstrumentationConfig)
+	if !ok {
+		return false
+	}
+	// we only care about new InstrumentationConfig objects
+	// the event will be triggered also when odiglet starts, and the instrumentation config is created
+	// in controller-runtime cache
+	// checking the RuntimeDetailsByContainer map should filter cases where we do not want to trigger the reconcile
+	// for workload which has already been reconciled previously
+	return len(obj.Status.RuntimeDetailsByContainer) == 0
+}
+
+func (p *instrumentationConfigPredicate) Update(e event.UpdateEvent) bool {
+	return false
+}
+
+func (p *instrumentationConfigPredicate) Delete(e event.DeleteEvent) bool {
+	return false
+}
+
+func (p *instrumentationConfigPredicate) Generic(e event.GenericEvent) bool {
+	return false
+}
+
+type InstrumentationConfigReconciler struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	Clientset *kubernetes.Clientset
+}
+
+func (r *InstrumentationConfigReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	logger := log.FromContext(ctx)
+
+	var instrumentationConfig odigosv1.InstrumentationConfig
+	err := r.Get(ctx, request.NamespacedName, &instrumentationConfig)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if len(instrumentationConfig.OwnerReferences) != 1 {
+		return reconcile.Result{}, fmt.Errorf("InstrumentationConfig %s/%s has %d owner references, expected 1", instrumentationConfig.Namespace, instrumentationConfig.Name, len(instrumentationConfig.OwnerReferences))
+	}
+
+	workload, labels, err := getWorkloadAndLabelsfromOwner(ctx, r.Client, instrumentationConfig.Namespace, instrumentationConfig.OwnerReferences[0])
+	if err != nil {
+		logger.Error(err, "Failed to get workload and labels from owner")
+		return reconcile.Result{}, err
+	}
+
+	pods, err := kubeutils.GetRunningPods(ctx, labels, workload.GetNamespace(), r.Client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// we need to apply runtime detection for just one pod with newer generation
+	var podForRuntimeDetails *corev1.Pod
+	selectedPodGeneration := int64(0)
+	for _, pod := range pods {
+		podGeneration, err := GetPodGeneration(ctx, r.Clientset, &pod)
+		if err != nil {
+			logger.Error(err, "Failed to get pod generation")
+			return reconcile.Result{}, err
+		}
+		if podGeneration == 0 {
+			// 0 means the pod is not relevant for runtime detection
+			continue
+		}
+
+		if podGeneration > instrumentationConfig.Status.ObservedWorkloadGeneration {
+			if podGeneration > selectedPodGeneration {
```

Comment:
`selectedPodGeneration` is going to be the latest one, right? If so, worth renaming it.

Reply:
This is the generation of the pod we selected for `podForRuntimeDetails`.
It will be the higher generation we found, not the last one

Reply:
renamed the pod variable to "selectedPodForInspection"